ledger/test-common/src/main/scala/com.digitalasset.platform.testing/SizeBoundObserver.scala

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.platform.testing
+
+import io.grpc.Context
+import io.grpc.stub.StreamObserver
+
+import scala.concurrent.{Future, Promise}
+
+object SizeBoundObserver {
+
+  def apply[A](n: Int)(attach: StreamObserver[A] => Unit): Future[Vector[A]] = {
+    if (n < 1) {
+      Future.failed(
+        new IllegalArgumentException(
+          s"Invalid argument $n, `take` requires a strictly positive integer as an argument"))
+    } else {
+      val observer = new SizeBoundObserver[A](n)
+      attach(observer)
+      observer.result
+    }
+  }
+
+}
+
+final class SizeBoundObserver[A](cap: Int) extends StreamObserver[A] {
+
+  private[this] val promise = Promise[Vector[A]]()
+  private[this] val items = Vector.newBuilder[A]
+
+  val result: Future[Vector[A]] = promise.future
+  var counter = 0
```

Comment:
I think this should be marked as `@volatile`. I assume that gRPC won't guarantee to always run the stream observer on a single thread?

Reply:
We're synchronizing all operations on `promise`. Shouldn't that be enough?

Reply:
You are right of course.